src/table/utils.cc

```diff
@@ -0,0 +1,164 @@
+        recordPtr->set(key, *valuePtr);
+        ++valuePtr;
+    }
+}
+
+template <typename Record, typename Key, typename Value>
+void setValues(SortedCatalogT<Record> &catalog, Key const &key, std::vector<Value> const &valueList) {
+    auto valuePtr = valueList.cbegin();
+    for (auto &record : catalog) {
+        record.set(key, *valuePtr);
+        ++valuePtr;
+    }
+}
+
+}  // namespace
+
+template <typename ReferenceCollection>
+void updateRefCentroids(geom::SkyWcs const &wcs, ReferenceCollection &refList) {
+    if (refList.size() < 1) {
+        return;
+    }
+    auto const schema = getSchema(refList);
+    CoordKey const coordKey(schema["coord"]);
+    Point2DKey const centroidKey(schema["centroid"]);
+    std::vector<coord::IcrsCoord> const skyList = getValues<coord::IcrsCoord>(refList, coordKey);
+    std::vector<geom::Point2D> const pixelList = wcs.skyToPixel(skyList);
+    setValues(refList, centroidKey, pixelList);
+}
+
+template <typename SourceCollection>
+void updateSourceCoords(geom::SkyWcs const &wcs, SourceCollection &sourceList) {
+    if (sourceList.size() < 1) {
+        return;
+    }
+    auto const schema = getSchema(sourceList);
+    Point2DKey const centroidKey(schema["slot_Centroid"]);
+    CoordKey const coordKey(schema["coord"]);
+    std::vector<geom::Point2D> pixelList = getValues<geom::Point2D>(sourceList, centroidKey);
+    std::vector<coord::IcrsCoord> const skyList = wcs.pixelToSky(pixelList);
+    setValues(sourceList, coordKey, skyList);
+}
+
+/// @cond
+
+#define INSTANTIATE_GET_SCHEMA(RECORD)                                             \
+    namespace {                                                                    \
+    template Schema getSchema(SortedCatalogT<const RECORD> const &);               \
+    template Schema getSchema(std::vector<std::shared_ptr<const RECORD>> const &); \
+    }
+#define INSTANTIATE_GET_SET_VALUES(RECORD, KEY, VALUE)                                               \
+    namespace {                                                                                      \
+    template std::vector<VALUE> getValues<VALUE>(SortedCatalogT<const RECORD> const &, KEY const &); \
+    template std::vector<VALUE> getValues<VALUE>(std::vector<std::shared_ptr<const RECORD>> const &, \
+                                                 KEY const &);                                       \
+    template void setValues(SortedCatalogT<RECORD> &, KEY const &, std::vector<VALUE> const &);      \
+    template void setValues(std::vector<std::shared_ptr<RECORD>> &, KEY const &,                     \
+                            std::vector<VALUE> const &);                                             \
+    }
+#define INSTANTIATE_HELPERS(RECORD)                                 \
+    INSTANTIATE_GET_SCHEMA(RECORD);                                 \
+    INSTANTIATE_GET_SET_VALUES(RECORD, CoordKey, coord::IcrsCoord); \
+    INSTANTIATE_GET_SET_VALUES(RECORD, Point2DKey, geom::Point2D);
```

Comment:
You are right. initially I found I had to do it to avoid linker errors. Either I messed up earlier or adding and using `FunctorKey::Value` somehow removed the need.